File: mix_cpp/src/line_follower.cpp
#include"rclcpp/rclcpp.hpp"
#include"rclcpp/node.hpp"
#include"sensor_msgs/msg/range.hpp"
#include"geometry_msgs/msg/twist.hpp"
#include"std_msgs/msg/float64.hpp"


const double MAX_RANGE = 0.5;
const double MAX_SPEED = 10.2;


// Versione in c++ Del segui linea per il robot tiago_base....


using std::placeholders::_1;
class LineFollower : public rclcpp::Node{

    public:
    rclcpp::Subscription<std_msgs::msg::Float64>::SharedPtr sub_ir_left;
    rclcpp::Subscription<std_msgs::msg::Float64>::SharedPtr sub_ir_right;
    rclcpp::Subscription<sensor_msgs::msg::Range>::SharedPtr sub_sonar_middle;

    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr pub_vel;


    double right_sensor_value;
    double left_sensor_value;
    double  middle_sensor_value;
    
    bool left_libero;
    bool right_libero;
    bool middle_libero;

    bool inizio;
    bool fine;


    LineFollower(): Node("LineFollower_cpp"){ 

        // Inizializzazione

        right_sensor_value = 0.0;
        left_sensor_value = 0.0;
        middle_sensor_value = 0.0;

        left_libero = false;
        right_libero = false;
        middle_libero = false;


        inizio = true;
        fine = false;

        // Costruttore in cui dichiare i miei subscriber alla lettura dei sensori

        sub_ir_left = this->create_subscription<std_msgs::msg::Float64>(
            "tiago_ir_left", 10, std::bind(&LineFollower::left_sensor_callback, this, _1));

        sub_ir_right = this->create_subscription<std_msgs::msg::Float64>(
            "tiago_ir_right", 10 , std::bind(&LineFollower::right_sensor_callback, this, _1));

        sub_sonar_middle = this->create_subscription<sensor_msgs::msg::Range>(
            "middle_sensor",10,std::bind(&LineFollower::middle_sensor_callback,this,_1));

        // Definisco qui il mio publisher 

        pub_vel = this->create_publisher<geometry_msgs::msg::Twist>("tiago_cmd_vel",10);

        // Non ho una funzione  in quanto devo solo pubblicare dei dati al topic tiago_cmd_vel

    }

    private:

    void left_sensor_callback(std_msgs::msg::Float64::SharedPtr msg) {

        left_sensor_value = msg.get()->data;
        RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "sensore sinistro");

        if(left_sensor_value <5.0)

            left_libero = false;
        
        else
            left_libero = true;
    
    }

    void right_sensor_callback(std_msgs::msg::Float64::SharedPtr msg){
          
        right_sensor_value = msg.get()->data;
        RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "sensore destro");

        if(right_sensor_value < 5.0)

            right_libero = false;

        else

            right_libero = true;

    }

    void middle_sensor_callback( sensor_msgs::msg::Range::SharedPtr msg){

        middle_sensor_value = msg.get()->range;
        RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "sensore di mezzo");

        // Oggetto da mandare al topic cmd_vel_per il settaggio delle velocità

        auto request = geometry_msgs::msg::Twist();

        if(inizio == true){

            RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Inizio = True");

            if(left_libero == false && right_libero == false){

                RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "inizio entrambi i sensori sulla riga nera");

                request.linear.x = -1.0;
                request.angular.z = 0.7;

                // Devo pubblicare il messaggio ... definire la funzione 
                
                pub_vel->publish(request);
            }
            else

                inizio = false;
    
        }

        else{

            RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Inizio = False");

            if(left_libero == true && right_libero == true && fine == false){

                request.linear.x = -0.5;
                request.angular.z = 0.0;
                pub_vel->publish(request);

            }

            else if(left_libero == true && right_libero == false && fine == false){

                request.linear.x = -0.5;
                request.angular.z = 1.0;
                pub_vel->publish(request);

            }

            else if(left_libero == false && right_libero == true && fine == false){

                request.linear.x = -0.5;
                request.angular.z = -1.0;
                pub_vel->publish(request);

            }

            else if(left_libero == false && right_libero == false && fine == false){

                request.linear.x = 0.0;
                request.angular.z = 0.0;
                pub_vel->publish(request);
                fine = true;

            }

        }

    }

};

int main(int argc, char **argv){

    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<LineFollower>());  
    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Inizio delle operazione di Line follower....");
    rclcpp::shutdown();

    return 0;



}

File: test_finale/test_finale/line_follower.py
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Range
from geometry_msgs.msg import Twist
from std_msgs.msg import Float64
# Range è un messaggio di tipo float

# Sensore sul nero valore 0.026 --- 0.046 (?)

MAX_RANGE = 0.5
MAX_SPEED = 10.2


cmd = Twist()


class LineFollower(Node):
    
    # Mezzo e sinistra 
    
    def __init__(self):
        
        super().__init__('tiago_line_follower_avoidance')
        
        self.inizio = False
        self.fine = False

        # self.create_subscription(Float64, 'tiago_ir_left', self.__left_sensor_callback, 1) # Sensore a IR
        
        # self.create_subscription(Float64, 'tiago_ir_right', self.__right_sensor_callback, 1)    # Sensore a IR
        
        # self.create_subscription(Range, 'middle_sensor', self.__middle_sensor_callback, 1) #   Sensore per ostacoli davanti
        
        
        self.create_subscription(Float64, 'tiago_ir_left', self.leftIR_cb, 1) # Sensore a IR
        
        self.create_subscription(Float64, 'tiago_ir_right', self.rightIR_cb, 1)    # Sensore a IR
        
        self.create_subscription(Range, 'tiago_middle', self.middleSensor_cb, 1) #   Sensore per ostacoli davanti
        
        self.create_subscription(Float64, 'tiago_light_sensor', self.lightSensor_cb, 1) #   Sensore per ostacoli davanti

        self.__publisher = self.create_publisher(Twist, 'tiago_cmd_vel', 1)

        self.left_libero = False
        self.right_libero = False
        self.middle_libero = False
        
        # inizializzazione
        self.GS_LEFT = 0.0 
        self.GS_RIGHT = 0.0
        self.GS_middle = 100
        self.GS_LIGHT = 256
        
        
        # self.velocita = -1.0# Parametro di velocita del robot 
        # self.velocita_angolare = 0.0
        
    def LineFollowModule(self):
        
        self.get_logger().info('chiama delle funzione LineFollowModule')  
        
        if(self.GS_middle > 0.4 or self.GS_LIGHT> 10):
            
            print('Non è stato rilevato alcuno ostacolo')

            if(self.GS_LEFT >= 500 and self.GS_RIGHT >= 500):
                print('una delle due ruote è su olio....')
                
                
                if(self.GS_LEFT < 500 and self.GS_RIGHT >=500):
                    cmd.linear.x = -0.1
                    cmd.angular.x = -0.2
                    self.__publisher.publish(cmd)

                
                elif(self.GS_LEFT >= 500 and self.GS_RIGHT < 500):
                
                    cmd.linear.x = -0.1
                    cmd.angular.x = 0.2
                    self.__publisher.publish(cmd)

                
                elif(self.GS_LEFT >= 500 and self.GS_RIGHT >= 500):
                    
                    cmd.linear.x = -0.12
                    cmd.angular.x = -0.1
                    self.__publisher.publish(cmd)

    
            else:
            
                print('fase di non slittamento')
            
            
                if(self.GS_LEFT < 100 and self.GS_RIGHT >= 100):
                    
                    cmd.linear.x = -0.1
                    cmd.angular.x = -0.2
                    self.__publisher.publish(cmd)

                    
                elif(self.GS_LEFT >= 100 and self.GS_RIGHT < 100):
                    
                    cmd.linear.x = -0.1
                    cmd.angular.x = 0.2
                    self.__publisher.publish(cmd)

                    
                elif(self.GS_LEFT >= 100 and self.GS_RIGHT >= 100):
                        
                    cmd.linear.x = -0.12
                    cmd.angular.x = -0.1
                    self.__publisher.publish(cmd)
        
                elif(self.GS_LEFT < 100 and self.GS_RIGHT < 100):
                        
                    cmd.linear.x = 0.0
                    cmd.angular.x = 0.0
                    self.__publisher.publish(cmd)
        else:
            
            if(self.GS_LIGHT <= 10):
                print('illuminazione non va bene per segui linea ')
                
            if(self.GS_middle <= 0.4):
                print('trovato un ostacolo')
                
            cmd.linear.x = 0.0
            cmd.angular.x = 0.0
            self.__publisher.publish(cmd)

    # Sono 3 che leggono i valori dei sensori, non serve richiarmarli esplicitamente
           
    def rightIR_cb(self,msg):
        
        print('destro: ' + str(msg.data))
        self.GS_RIGHT = msg.data # Non prendo oggetto Float ma oggetto.data
        
    def leftIR_cb(self,msg):
        
        print('sinistro: ' + str(msg.data))
        
        self.GS_LEFT = msg.data # Non prendo oggetto Float ma oggetto.data
        
    def lightSensor_cb(self,msg):
        
        print('illuminazione: ' + str(msg.data))
        self.GS_LIGHT = msg.data
                
    def middleSensor_cb(self,msg):
        print('centrale: ' + str(msg.range))

        self.GS_middle = msg.range # Non prendo oggetto Range ma oggetto.range
        print('range: ' + str(self.GS_middle))
        self.LineFollowModule()

        self.fine = True
 
        
    #     if(self.GS_middle>0.2):
            
    #         print('Non è stato rilevato alcuno ostacolo')

    #         if(self.GS_LEFT < 100 and self.GS_RIGHT >= 100):
                
    #             cmd.linear.x = -0.1
    #             cmd.angular.x = -0.2
    #             self.__publisher.publish(cmd)

                
    #         elif(self.GS_LEFT >= 100 and self.GS_RIGHT < 100):
                
    #             cmd.linear.x = -0.1
    #             cmd.angular.x = 0.2
    #             self.__publisher.publish(cmd)

                
    #         elif(self.GS_LEFT >= 100 and self.GS_RIGHT >= 100):
                
    #             cmd.linear.x = -0.12
    #             cmd.angular.x = -0.1
    #             self.__publisher.publish(cmd)

                
    #         elif(self.GS_LEFT < 100 and self.GS_RIGHT < 100): 
                
    #             cmd.linear.x = 0.0
    #             cmd.angular.x = 0.0
    #             self.__publisher.publish(cmd)
  
    #     else:
            
            
    #         if(self.GS_middle <= 0.2):
                
    #             self.get_logger().info('chiama delle funzione per apertura e chiusura della porta')   

                
    #         cmd.linear.x = -0.0
    #         cmd.angular.x = 0.0
    #         self.__publisher.publish(cmd)
    
    # # Sono 3 che leggono i valori dei sensori, non serve richiarmarli esplicitamente
           
    def rightIR_cb(self,msg):
        
        print('aggiornamento destro: ' + str(msg.data))
        self.GS_RIGHT = msg.data # Non prendo oggetto Float ma oggetto.data
        
    def leftIR_cb(self,msg):
        
        print('aggiornamento sinistro: ' + str(msg.data))
        
        self.GS_LEFT = msg.data # Non prendo oggetto Float ma oggetto.data
        
    def middleSensor_cb(self,msg):
        print('aggiornamento centrale: ' + str(msg.range))

        self.GS_middle = msg.range # Non prendo oggetto Range ma oggetto.range
        print('range: ' + str(self.GS_middle))
        self.LineFollowModule()

        self.fine = True
        

def main(args=None):
    
    rclpy.init(args=args)
    follower = LineFollower()
        
    rclpy.spin(follower)
    


    follower.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()